build_tools/packaging/linux/upload_package_repo.py

```diff
@@ -74,7 +83,232 @@ def generate_indexes_recursive(root):
+
+            # Step 1: Download existing repodata from S3 (small files)
+            old_repodata_dir = old_repo_dir / "repodata"
+            old_repodata_dir.mkdir(parents=True, exist_ok=True)
+
+            print("Downloading existing repository metadata from S3...")
+            repodata_files = []
+            try:
+                paginator = s3.get_paginator("list_objects_v2")
+                for page in paginator.paginate(
+                    Bucket=bucket, Prefix=f"{prefix}/x86_64/repodata/"
+                ):
+                    if "Contents" not in page:
+                        continue
+                    for obj in page["Contents"]:
+                        key = obj["Key"]
+                        filename = Path(key).name
+                        local_file = old_repodata_dir / filename
+                        s3.download_file(bucket, key, str(local_file))
+                        repodata_files.append(filename)
+                        print(f"  Downloaded: {filename}")
+            except Exception as e:
+                print(f"Note: No existing repodata found (new repo?): {e}")
+
+            # Step 2: Generate repodata for NEW packages only (from local dir)
+            local_arch_dir = Path(local_package_dir) / "x86_64"
+            if local_arch_dir.exists() and list(local_arch_dir.glob("*.rpm")):
+                print("Generating metadata for new packages...")
+                # Copy new RPMs to temp dir
+                new_arch_dir = new_repo_dir / "x86_64"
+                new_arch_dir.mkdir(parents=True, exist_ok=True)
+                for rpm_file in local_arch_dir.glob("*.rpm"):
+                    shutil.copy2(rpm_file, new_arch_dir / rpm_file.name)
+
+                # Generate repodata for new packages
+                run_command("createrepo_c .", cwd=str(new_arch_dir))
+                print("✅ Generated metadata for new packages")
+            else:
+                print("No new RPM packages to process")
+                return
+
+            # Step 3: Merge repositories using mergerepo_c (no need to download all RPMs!)
+            merged_arch_dir = merged_repo_dir / "x86_64"
+            merged_arch_dir.mkdir(parents=True, exist_ok=True)
+
+            if repodata_files:  # If we have existing metadata
+                print("Merging old and new repository metadata...")
+                # mergerepo_c merges repodata without needing actual RPM files!
+                run_command(
+                    f'mergerepo_c --repo "{old_repo_dir}" --repo "{new_repo_dir / "x86_64"}" --outputdir "{merged_arch_dir}"',
+                    cwd=str(temp_path),
+                )
+                print("✅ Merged repository metadata")
+            else:  # First upload, no existing metadata
+                print("First upload - using new repository metadata")
+                shutil.copytree(
+                    new_repo_dir / "x86_64" / "repodata", merged_arch_dir / "repodata"
```

Comment:
shutil.copytree supports( dirs_exist_ok=True) 
This avoids FileExistsError if the directory was created earlier.
Pls note shutil.copy2 doesn't supports the same